File: backend/javascript/test_api.js
#!/usr/bin/env node
/**
 * Test suite for JavaScript REPL backend
 * Tests the core functionality: execution, persistence, and reset
 */

const http = require('http')
const crypto = require('crypto')

const BASE_URL = 'http://localhost:8000'
const TEST_SESSION_ID = crypto.randomUUID()

// Helper function to make HTTP requests
function makeRequest(method, path, data = null) {
  return new Promise((resolve, reject) => {
    const url = new URL(BASE_URL + path)
    const options = {
      hostname: url.hostname,
      port: url.port,
      path: url.pathname,
      method: method,
      headers: {
        'Content-Type': 'application/json'
      }
    }

    const req = http.request(options, (res) => {
      let body = ''
      res.on('data', chunk => body += chunk)
      res.on('end', () => {
        try {
          const response = {
            status: res.statusCode,
            data: JSON.parse(body)
          }
          resolve(response)
        } catch (e) {
          reject(new Error(`Failed to parse response: ${e.message}`))
        }
      })
    })

    req.on('error', reject)
    
    if (data) {
      req.write(JSON.stringify(data))
    }
    req.end()
  })
}

async function testHealthCheck() {
  const response = await makeRequest('GET', '/health')
  console.assert(response.status === 200, 'Health check status should be 200')
  console.assert(response.data.status === 'ok', 'Health status should be ok')
  console.assert(response.data.language === 'javascript', 'Language should be javascript')
  console.log('✓ Health check passed')
}

async function testSimpleExpression() {
  const response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { code: '2 + 2' })
  console.assert(response.status === 200, 'Execute status should be 200')
  console.assert(response.data.output === '4\n', 'Output should be 4')
  console.assert(response.data.error === null, 'Error should be null')
  console.log('✓ Simple expression evaluation passed')
}

async function testConsoleOutput() {
  const response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { code: 'console.log("Hello, JavaScript!")' })
  console.assert(response.status === 200, 'Execute status should be 200')
  console.assert(response.data.output === 'Hello, JavaScript!\n', 'Output should match')
  console.assert(response.data.error === null, 'Error should be null')
  console.log('✓ Console output capture passed')
}

async function testVariablePersistence() {
  // Set a variable
  let response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { code: 'let testVar = 42' })
  console.assert(response.status === 200, 'Execute status should be 200')
  
  // Use the variable in another execution
  response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { code: 'testVar * 2' })
  console.assert(response.status === 200, 'Execute status should be 200')
  console.assert(response.data.output === '84\n', 'Output should be 84')
  console.assert(response.data.error === null, 'Error should be null')
  console.log('✓ Variable persistence passed')
}

async function testErrorHandling() {
  const response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { code: 'undefinedVariable.property' })
  console.assert(response.status === 200, 'Execute status should be 200')
  console.assert(response.data.error !== null, 'Error should not be null')
  console.assert(response.data.error.includes('TypeError'), 'Error should be TypeError')
  console.log('✓ Error handling passed')
}

async function testResetFunctionality() {
  // Set a variable
  let response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { code: 'let resetTestVar = 999' })
  console.assert(response.status === 200, 'Execute status should be 200')
  
  // Reset the environment
  response = await makeRequest('POST', `/reset/${TEST_SESSION_ID}`)
  console.assert(response.status === 200, 'Reset status should be 200')
  console.assert(response.data.message === 'Context reset successfully', 'Reset message should match')
  
  // Try to access the variable (should be undefined)
  response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { code: 'typeof resetTestVar' })
  console.assert(response.status === 200, 'Execute status should be 200')
  console.assert(response.data.output === 'undefined\n', 'Variable should be undefined after reset')
  console.log('✓ Reset functionality passed')
}

async function testMultilineCode() {
  const code = `
function factorial(n) {
  if (n <= 1) return 1
  return n * factorial(n - 1)
}
console.log(factorial(5))
`
  const response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { code })
  console.assert(response.status === 200, 'Execute status should be 200')
  console.assert(response.data.output === '120\n', 'Output should be 120')
  console.assert(response.data.error === null, 'Error should be null')
  console.log('✓ Multiline code execution passed')
}

async function testObjectOutput() {
  const response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { 
    code: 'console.log({name: "test", value: 123})' 
  })
  console.assert(response.status === 200, 'Execute status should be 200')
  console.assert(response.data.output.includes('"name"'), 'Output should contain object properties')
  console.assert(response.data.output.includes('"test"'), 'Output should contain object values')
  console.log('✓ Object output serialization passed')
}

async function testAsyncCode() {
  const code = `
const promise = new Promise(resolve => resolve(42))
promise.then(val => console.log('Promise resolved:', val))
`
  const response = await makeRequest('POST', `/execute/${TEST_SESSION_ID}`, { code })
  console.assert(response.status === 200, 'Execute status should be 200')
  // Note: Promise won't resolve in sync execution, but shouldn't error
  console.assert(response.data.error === null, 'Error should be null')
  console.log('✓ Async code handling passed')
}

async function runAllTests() {
  console.log('\n=== JavaScript Backend Test Suite ===\n')
  
  const tests = [
    testHealthCheck,
    testSimpleExpression,
    testConsoleOutput,
    testVariablePersistence,
    testErrorHandling,
    testResetFunctionality,
    testMultilineCode,
    testObjectOutput,
    testAsyncCode
  ]
  
  let failed = 0
  for (const test of tests) {
    try {
      await test()
    } catch (error) {
      console.log(`✗ ${test.name} failed: ${error.message}`)
      if (error.code === 'ECONNREFUSED') {
        console.log('  Make sure the JavaScript backend is running on port 8000')
      }
      failed++
    }
  }
  
  console.log(`\n=== Results: ${tests.length - failed}/${tests.length} tests passed ===\n`)
  return failed === 0
}

// Run tests
runAllTests().then(success => {
  process.exit(success ? 0 : 1)
}).catch(error => {
  console.error('Test suite failed:', error)
  process.exit(1)
})

File: backend/ruby/test_api.rb
#!/usr/bin/env ruby
# Test suite for Ruby REPL backend
# Tests the core functionality: execution, persistence, and reset

require 'net/http'
require 'json'
require 'uri'
require 'securerandom'

BASE_URL = 'http://localhost:8000'
TEST_SESSION_ID = SecureRandom.uuid

class TestRunner
  def initialize
    @passed = 0
    @failed = 0
  end

  def make_request(method, path, data = nil)
    uri = URI.parse("#{BASE_URL}#{path}")
    
    case method
    when :get
      request = Net::HTTP::Get.new(uri)
    when :post
      request = Net::HTTP::Post.new(uri)
      request.content_type = 'application/json'
      request.body = data.to_json if data
    end

    response = Net::HTTP.start(uri.hostname, uri.port) do |http|
      http.request(request)
    end

    {
      status: response.code.to_i,
      body: JSON.parse(response.body)
    }
  rescue => e
    raise "Request failed: #{e.message}"
  end

  def assert(condition, message)
    unless condition
      raise AssertionError, message
    end
  end

  def test_health_check
    response = make_request(:get, '/health')
    assert(response[:status] == 200, 'Health check status should be 200')
    assert(response[:body]['status'] == 'ok', 'Health status should be ok')
    assert(response[:body]['language'] == 'ruby', 'Language should be ruby')
    puts '✓ Health check passed'
  end

  def test_simple_expression
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: '2 + 2' })
    assert(response[:status] == 200, 'Execute status should be 200')
    assert(response[:body]['output'] == "4\n", 'Output should be 4')
    assert(response[:body]['error'].nil?, 'Error should be nil')
    puts '✓ Simple expression evaluation passed'
  end

  def test_puts_output
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: 'puts "Hello, Ruby!"' })
    assert(response[:status] == 200, 'Execute status should be 200')
    assert(response[:body]['output'] == "Hello, Ruby!\n", 'Output should match')
    assert(response[:body]['error'].nil?, 'Error should be nil')
    puts '✓ Puts output capture passed'
  end

  def test_variable_persistence
    # Set a variable
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: '@test_var = 42' })
    assert(response[:status] == 200, 'Execute status should be 200')
    
    # Use the variable in another execution
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: '@test_var * 2' })
    assert(response[:status] == 200, 'Execute status should be 200')
    assert(response[:body]['output'] == "84\n", 'Output should be 84')
    assert(response[:body]['error'].nil?, 'Error should be nil')
    puts '✓ Variable persistence passed'
  end

  def test_error_handling
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: '1 / 0' })
    assert(response[:status] == 200, 'Execute status should be 200')
    assert(!response[:body]['error'].nil?, 'Error should not be nil')
    assert(response[:body]['error'].include?('ZeroDivisionError'), 'Error should be ZeroDivisionError')
    puts '✓ Error handling passed'
  end

  def test_reset_functionality
    # Set a variable
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: '@reset_test_var = 999' })
    assert(response[:status] == 200, 'Execute status should be 200')
    
    # Reset the environment
    response = make_request(:post, "/reset/#{TEST_SESSION_ID}")
    assert(response[:status] == 200, 'Reset status should be 200')
    assert(response[:body]['message'] == 'Environment reset successfully', 'Reset message should match')
    
    # Try to access the variable (should be nil)
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: '@reset_test_var' })
    assert(response[:status] == 200, 'Execute status should be 200')
    # In Ruby, undefined instance variables return nil
    assert(response[:body]['output'] == "" || response[:body]['output'] == "\n" || response[:body]['output'] == "nil\n", 
           'Variable should be nil after reset')
    puts '✓ Reset functionality passed'
  end

  def test_multiline_code
    code = <<~RUBY
      def factorial(n)
        return 1 if n <= 1
        n * factorial(n - 1)
      end
      puts factorial(5)
    RUBY
    
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: code })
    assert(response[:status] == 200, 'Execute status should be 200')
    assert(response[:body]['output'] == "120\n", 'Output should be 120')
    assert(response[:body]['error'].nil?, 'Error should be nil')
    puts '✓ Multiline code execution passed'
  end

  def test_class_definition
    # Define a class
    code1 = <<~RUBY
      class TestPerson
        attr_accessor :name, :age
        
        def initialize(name, age)
          @name = name
          @age = age
        end
        
        def greeting
          "Hello, I'm \#{@name} and I'm \#{@age} years old"
        end
      end
    RUBY
    
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: code1 })
    assert(response[:status] == 200, 'Execute status should be 200')
    
    # Use the class
    code2 = 'person = TestPerson.new("Alice", 30); puts person.greeting'
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: code2 })
    assert(response[:status] == 200, 'Execute status should be 200')
    assert(response[:body]['output'] == "Hello, I'm Alice and I'm 30 years old\n", 'Output should match')
    puts '✓ Class definition and usage passed'
  end

  def test_array_operations
    code = '[1, 2, 3, 4, 5].map { |n| n * 2 }.sum'
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: code })
    assert(response[:status] == 200, 'Execute status should be 200')
    assert(response[:body]['output'] == "30\n", 'Output should be 30')
    assert(response[:body]['error'].nil?, 'Error should be nil')
    puts '✓ Array operations passed'
  end

  def test_string_interpolation
    # Set a variable
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: '@name = "Ruby"' })
    assert(response[:status] == 200, 'Execute status should be 200')
    
    # Use string interpolation
    response = make_request(:post, "/execute/#{TEST_SESSION_ID}", { code: 'puts "Hello from #{@name}!"' })
    assert(response[:status] == 200, 'Execute status should be 200')
    assert(response[:body]['output'] == "Hello from Ruby!\n", 'Output should match')
    puts '✓ String interpolation passed'
  end

  def run_all_tests
    puts "\n=== Ruby Backend Test Suite ===\n"
    
    tests = [
      :test_health_check,
      :test_simple_expression,
      :test_puts_output,
      :test_variable_persistence,
      :test_error_handling,
      :test_reset_functionality,
      :test_multiline_code,
      :test_class_definition,
      :test_array_operations,
      :test_string_interpolation
    ]
    
    tests.each do |test|
      begin
        send(test)
        @passed += 1
      rescue => e
        puts "✗ #{test} failed: #{e.message}"
        if e.message.include?('Connection refused')
          puts '  Make sure the Ruby backend is running on port 8000'
        end
        @failed += 1
      end
    end
    
    puts "\n=== Results: #{@passed}/#{tests.length} tests passed ===\n"
    @failed == 0
  end
end

# Custom error class for assertions
class AssertionError < StandardError; end

# Run tests
if __FILE__ == $0
  runner = TestRunner.new
  success = runner.run_all_tests
  exit(success ? 0 : 1)
end